byterun/caml/weak.h

```diff
@@ -18,10 +18,122 @@
+/** Return 1 if the key in the ephemeron at the given offset is set.
+    Otherwise 0. When returning 1, set [*key] to a shallow copy of the
+    key. This function allocates.
+
+    The value [eph] must be an ephemeron and [offset] a valid key
+    offset.
+*/
+
+CAMLextern void caml_ephemeron_blit_key(value eph1, mlsize_t off1,
+                                        value eph2, mlsize_t off2,
+                                        mlsize_t len);
+/** Fill the given range of keys of [eph2] with the given range of
+    keys of [eph1]. Contrary to using caml_ephemeron_get_key followed
+    by caml_ephemeron_set_key or caml_ephemeron_unset_key, this
+    function does not prevent the incremental GC from erasing the
+    value in its current cycle. The value [eph1] (resp. [eph2]) must
+    be an ephemeron and the offsets between [off1] and [off1+len]
+    (resp. between [off2] and [off2+offset]) must be valid keys of
+    [eph1] (resp. [eph2]).
+*/
+
+CAMLextern int caml_ephemeron_data_is_set(value eph);
+/** Return 1 if the data in the ephemeron is set.
+    Otherwise 0. The value [eph] must be an ephemeron.
+*/
+
+CAMLextern void caml_ephemeron_set_data(value eph, value k);
+/** Set the data of the given ephemeron [eph] to the given value
+    [k]. The value [eph] must be an ephemeron and [k] a block.
+*/
+
+CAMLextern void caml_ephemeron_unset_data(value eph);
+/** Unset the data of the given ephemeron. The value [eph] must be an
+    ephemeron.
+*/
+
+CAMLextern int caml_ephemeron_get_data(value eph, value *data);
+/** Return 1 if the data in the ephemeron at the given offset is set.
+    Otherwise 0. When returning 1, set [*data] to the pointed value.
+
+    The value [eph] must be an ephemeron and [offset] a valid key
+    offset.
+*/
+
+CAMLextern int caml_ephemeron_get_data_copy(value eph, value *data);
+/** Return 1 if the data in the ephemeron at the given offset is set.
+    Otherwise 0. When returning 1, set [*data] to a shallow copy of
+    the data. This function allocates.
+
+    The value [eph] must be an ephemeron and [offset] a valid key
+    offset.
+*/
+
+CAMLextern void caml_ephemeron_blit_data(value eph1, value eph2);
+/** Sets the data of [eph2] to be the same as the data of [eph1].
+    Contrary to using caml_ephemeron_get_data followed by
+    caml_ephemeron_set_data or caml_ephemeron_unset_data, this
+    function does not prevent the incremental GC from erasing the
+    value in its current cycle. The values [eph1] and [eph2] must be
+    ephemerons.
+*/
+
+
+#define caml_weak_array_length caml_ephemeron_num_keys
+#define caml_weak_array_create caml_ephemeron_create
+#define caml_weak_array_check caml_ephemeron_key_is_set
+#define caml_weak_array_unset caml_ephemeron_unset_key
+#define caml_weak_array_set caml_ephemeron_set_key
+#define caml_weak_array_get caml_ephemeron_get_key
+#define caml_weak_array_get_copy caml_ephemeron_get_key_copy
+#define caml_weak_array_blit caml_ephemeron_blit_key
```

Comment:
Minor detail: these aliases aren't very helpful for users of this API, maybe just making the arguments explicit would be clearer?
```c
#define caml_weak_array_get(array, offset, result) \
    caml_ephemeron_get_key(array, offset, result)
```
I am a bit uncomfortable with this API. It was easier for me to write my code as if I was working with ephemeron, then rewrite with the `weak_array` subset later because I just needed this part.
To me it seems this conflate weak arrays and ephemerons, while the fact that they are represented the same way is an implementation detail of the runtime.

Reply:
Weak arrays are now mostly deprecated because they are superseded by ephemerons.

Reply:
I have no opinion about whether `ephemeron.h` is more appropriate than `weak.h`.
Maybe the `caml_weak_...` variants of the functions are not necessary if weak arrays are deprecated (as far as C api is concerned)?
Anyway this is just a detail and I am happy with the PR as it is.